File: tools/openclinxr/asset-pipeline/makeclothes/licence_precedence.py
"""Licence precedence for the MakeHuman bake gates (OPERATOR RULING 2026-09-17,
REFINED by the operator the same day).

Mirrors tools/openclinxr/asset-pipeline/makeclothes/licence-precedence.ts: the
MORE PERMISSIVE of the catalogue listing and the asset's own file line governs
(ranks in licence_permissiveness_rank). New operator line (verbatim): "Review the
assets with their listing page - is the listing page more permissive? If so record that
as the license instead of the license embedded into the asset as many just leave the
default license." Earlier the same day: "go with what site links say (CC0 over AGPLv3)".
Stdlib only.
"""

import json
import pathlib
import re

REPO_ROOT = pathlib.Path(__file__).resolve().parents[4]
CATALOGUE_SNAPSHOT_PATH = (
    REPO_ROOT / "tools/openclinxr/asset-pipeline/makeclothes/makehuman-catalogue-snapshot.json"
)

_CATALOGUE_PACKS = None


def load_catalogue_packs(path=None):
    """Load the committed catalogue snapshot packs (cached; no network)."""
    global _CATALOGUE_PACKS
    if path is None and _CATALOGUE_PACKS is not None:
        return _CATALOGUE_PACKS
    snap_path = pathlib.Path(path) if path is not None else CATALOGUE_SNAPSHOT_PATH
    try:
        packs = json.loads(snap_path.read_text(encoding="utf-8")).get("packs", {})
    except OSError:
        packs = {}
    if path is None:
        _CATALOGUE_PACKS = packs
    return packs


def normalise_licence_token(token):
    """Normalise for conflict comparison so CC0/CC-0 spellings compare equal."""
    return re.sub(r"[\s_-]+", "", str(token).strip().lower())


def licence_permissiveness_rank(token):
    """Permissiveness rank for the more-permissive-wins comparison (redistribution
    freedom: CC0 imposes no conditions; CC-BY imposes attribution only; copyleft
    imposes source obligations; an unrecognised token imposes unknown obligations).
    CC0 / public domain 3, CC-BY family 2, AGPL/GPL copyleft 1, unrecognised 0.
    Silence has no rank — the callers treat it as "not a candidate", never a winner.
    """
    t = str(token).strip()
    if re.search(r"cc\s*[-_ ]?0\b", t, re.I) or re.search(r"public\s+domain", t, re.I):
        return 3
    if re.search(r"cc[\s_-]*by", t, re.I):
        return 2
    if re.search(r"agpl|gpl|copyleft", t, re.I):
        return 1
    return 0


def catalogue_verdict(pack_slug, packs=None):
    """Catalogue state for a pack slug: listed, conflict, or silent.

    Returns (status, entry) where status is one of "listed", "conflict",
    "silent". Conflict means more than one distinct licence value across the
    entry's `licence` + `licences` fields (normalised for spelling).
    """
    table = packs if packs is not None else load_catalogue_packs()
    entry = table.get(pack_slug) if pack_slug else None
    if not entry:
        return "silent", None
    candidates = [entry.get("licence")] + list(entry.get("licences") or [])
    distinct = {normalise_licence_token(c) for c in candidates if c}
    if len(distinct) > 1:
        return "conflict", entry
    return "listed", entry


def resolve_licence_precedence(pack_slug, file_declared, file_permitted, packs=None,
                               file_attribution_required=False, file_refusal_reason=None):
    """One precedence decision for every bake gate.

    pack_slug: catalogue pack (e.g. "hair01"), or None when unlisted.
    file_declared: raw licence line from the asset's own file, or None when silent.
    file_permitted: the file side's own verdict (True only if the header clears alone).
    Listed catalogue + declared file -> the MORE PERMISSIVE side governs
    (catalogue rank > file rank -> catalogue, with the file's declared line recorded
    as overridden; catalogue rank <= file rank -> the file's own verdict decides,
    via "file"). A silent file is never a "declared" side, so a listed catalogue
    always governs it.
    Returns a dict with permitted, via, licence, refusal_reason,
    overridden_file_licence.
    """
    table = packs if packs is not None else load_catalogue_packs()
    status, entry = catalogue_verdict(pack_slug, table)
    if status == "conflict":
        assert entry is not None  # conflict implies an entry; narrows the Optional for checkers
        candidates = [entry.get("licence")] + list(entry.get("licences") or [])
        names = sorted({str(c).strip() for c in candidates if str(c).strip()})
        return {
            "permitted": False,
            "via": "none",
            "licence": None,
            "refusal_reason": (
                "catalogue lists conflicting licences for pack "
                f"{pack_slug or '(none)'}: {' vs '.join(names)}"
                " — refused without consulting the file"
            ),
            "overridden_file_licence": None,
            "attribution_required": False,
        }
    if status == "listed":
        assert entry is not None  # listed implies an entry; narrows the Optional for checkers
        catalogue_licence = entry.get("licence")
        catalogue_is_by = bool(re.search(r"cc[\s_-]*by", str(catalogue_licence or ""), re.I))
        catalogue_rank = licence_permissiveness_rank(catalogue_licence)
        file_rank = licence_permissiveness_rank(file_declared) if file_declared else None
        if file_rank is None or catalogue_rank > file_rank:
            # A catalogue listing is a grant we may rely on only when the grant
            # itself clears the CC0/CC-BY bar (rank >= 2); otherwise refuse
            # without consulting the file (it already lost the comparison).
            if catalogue_rank < 2:
                return {
                    "permitted": False,
                    "via": "none",
                    "licence": None,
                    "refusal_reason": (
                        f"catalogue licence for pack {pack_slug or '(none)'} "
                        f"is not CC0/CC-BY: {catalogue_licence} — refused"
                    ),
                    "overridden_file_licence": None,
                    "attribution_required": False,
                }
            return {
                "permitted": True,
                "via": "catalogue",
                "licence": catalogue_licence,
                "refusal_reason": None,
                "overridden_file_licence": file_declared,
                "attribution_required": catalogue_is_by or bool(file_attribution_required),
            }
        # Tie below the CC0/CC-BY bar (e.g. garbled catalogue + garbled file):
        # neither side is a grant we may rely on — refuse without letting the
        # file decide. A file that clears the bar on its own still governs.
        if catalogue_rank == file_rank and catalogue_rank < 2:
            return {
                "permitted": False,
                "via": "none",
                "licence": None,
                "refusal_reason": (
                    f"catalogue licence for pack {pack_slug or '(none)'} "
                    f"is not CC0/CC-BY: {catalogue_licence} — refused"
                ),
                "overridden_file_licence": None,
                "attribution_required": False,
            }
        return {
            "permitted": bool(file_permitted),
            "via": "file",
            "licence": file_declared,
            "refusal_reason": file_refusal_reason,
            "overridden_file_licence": None,
            "attribution_required": bool(file_attribution_required),
        }
    if file_declared:
        return {
            "permitted": bool(file_permitted),
            "via": "file",
            "licence": file_declared,
            "refusal_reason": file_refusal_reason,
            "overridden_file_licence": None,
            "attribution_required": bool(file_attribution_required),
        }
    return {
        "permitted": False,
        "via": "none",
        "licence": None,
        "refusal_reason": (
            file_refusal_reason
            or "no licence in the file and no catalogue listing — unspecified is a refusal"
        ),
        "overridden_file_licence": None,
        "attribution_required": False,
    }

File: tools/openclinxr/asset-pipeline/makeclothes/licence-precedence.ts
/**
 * Licence precedence for the MakeHuman bake gates (OPERATOR RULING 2026-09-17,
 * REFINED by the operator the same day).
 *
 * New operator line (verbatim): "Review the assets with their listing page - is the
 * listing page more permissive? If so record that as the license instead of the license
 * embedded into the asset as many just leave the default license." Earlier the same day:
 * "go with what site links say (CC0 over AGPLv3)".
 *
 * So the catalogue listing no longer wins ALWAYS: the MORE PERMISSIVE of the catalogue
 * listing and the asset's own file line governs (ranks in licencePermissivenessRank).
 * This is the skins01 case: catalogue CC0 (rank 3) beats file AGPLv3 (rank 1). A listed
 * CC-BY catalogue over a file CC0 keeps the file CC0; equal ranks keep the file; a
 * garbled/unrecognised token (rank 0) never beats anything, including a copyleft file line.
 *
 * claimScope: precedence decision only; file-token parsing stays in the callers.
 * notEvidenceFor: per-asset grants outside the catalogue, user-contributed licences.
 */

import { catalogueEntryForPack, type CatalogueEntry } from "./makehuman-catalogue.js";

export type FileLicenceVerdict = {
  /** Raw licence line from the asset's own file, or null when silent. */
  declared: string | null;
  /** The file side's own verdict: permitted only when the header clears on its own. */
  permitted: boolean;
  /** True when the header declared a CC-BY family licence needing attribution. */
  attributionRequired: boolean;
  /** File-side refusal reason, or null when the file side permits. */
  refusalReason: string | null;
};

export type LicencePrecedenceResult = {
  permitted: boolean;
  via: "catalogue" | "file" | "none";
  licence: string | null;
  attributionRequired: boolean;
  refusalReason: string | null;
  /** The file's declared line when the catalogue governed and the file declared something. */
  overriddenFileLicence: string | null;
  catalogueEntry: CatalogueEntry | null;
};

export type LicencePrecedenceInput = {
  packSlug: string | null;
  file: FileLicenceVerdict;
  packs?: Record<string, CatalogueEntry>;
};

/**
 * One precedence decision for every bake gate.
 *
 * (a) Catalogue listing exists and is unambiguous -> compare permissiveness: the more
 *     permissive side governs (catalogue rank > file rank -> catalogue, with the file's
 *     declared line recorded as overridden; catalogue rank <= file rank -> the file's
 *     own verdict decides, via "file"). A silent file is never a "declared" side, so a
 *     listed catalogue always governs it.
 * (b) Catalogue silent for that pack -> the file's own verdict decides.
 * (c) Catalogue silent AND file silent -> refused.
 * (d) Catalogue lists conflicting licences for the same pack -> refused, naming
 *     the conflicting values; the file never rescues a conflict.
 *
 * Attribution: when the catalogue governs, attributionRequired = (catalogue
 * licence is CC-BY) OR (the file itself declared a CC-BY family licence).
 */
export function resolveLicencePrecedence(input: LicencePrecedenceInput): LicencePrecedenceResult {
  const { packSlug, file } = input;
  const entry =
    input.packs !== undefined
      ? ((packSlug ? input.packs[packSlug] : undefined) ?? null)
      : catalogueEntryForPack(packSlug);
  if (entry) {
    const candidates = [entry.licence, ...(entry.licences ?? [])];
    const distinct = [...new Set(candidates.map(normaliseLicenceToken))];
    if (distinct.length > 1) {
      const names = [...new Set(candidates.map((c) => String(c).trim()).filter(Boolean))].sort();
      return {
        permitted: false,
        via: "none",
        licence: null,
        attributionRequired: false,
        refusalReason:
          `catalogue lists conflicting licences for pack ${packSlug ?? "(none)"}: ` +
          `${names.join(" vs ")} — refused without consulting the file`,
        overriddenFileLicence: null,
        catalogueEntry: entry,
      };
    }
    const catalogueLicence = entry.licence;
    const catalogueIsBy = /cc[\s_-]*by/i.test(catalogueLicence);
    const catalogueRank = licencePermissivenessRank(catalogueLicence);
    const fileRank = file.declared ? licencePermissivenessRank(file.declared) : null;
    if (fileRank === null || catalogueRank > fileRank) {
      // A catalogue listing is a grant we may rely on only when the grant itself
      // clears the CC0/CC-BY bar (rank >= 2); otherwise refuse without consulting the file.
      if (catalogueRank < 2) {
        return {
          permitted: false,
          via: "none",
          licence: null,
          attributionRequired: false,
          refusalReason:
            `catalogue licence for pack ${packSlug ?? "(none)"} is not CC0/CC-BY: ` +
            `${catalogueLicence} — refused`,
          overriddenFileLicence: null,
          catalogueEntry: entry,
        };
      }
      return {
        permitted: true,
        via: "catalogue",
        licence: catalogueLicence,
        attributionRequired: catalogueIsBy || file.attributionRequired,
        refusalReason: null,
        overriddenFileLicence: file.declared,
        catalogueEntry: entry,
      };
    }
    // Tie below the CC0/CC-BY bar (e.g. garbled catalogue + garbled file): neither
    // side is a grant we may rely on — refuse without letting the file decide.
    // A file that clears the bar on its own (rank >= 2) still governs.
    if (catalogueRank === fileRank && catalogueRank < 2) {
      return {
        permitted: false,
        via: "none",
        licence: null,
        attributionRequired: false,
        refusalReason:
          `catalogue licence for pack ${packSlug ?? "(none)"} is not CC0/CC-BY: ` +
          `${catalogueLicence} — refused`,
        overriddenFileLicence: null,
        catalogueEntry: entry,
      };
    }
    return {
      permitted: file.permitted,
      via: "file",
      licence: file.declared,
      attributionRequired: file.attributionRequired,
      refusalReason: file.refusalReason,
      overriddenFileLicence: null,
      catalogueEntry: entry,
    };
  }
  if (file.declared) {
    return {
      permitted: file.permitted,
      via: "file",
      licence: file.declared,
      attributionRequired: file.attributionRequired,
      refusalReason: file.refusalReason,
      overriddenFileLicence: null,
      catalogueEntry: null,
    };
  }
  return {
    permitted: false,
    via: "none",
    licence: null,
    attributionRequired: false,
    refusalReason: file.refusalReason ?? "no licence in the file and no catalogue listing — unspecified is a refusal",
    overriddenFileLicence: null,
    catalogueEntry: null,
  };
}

/** Normalise for conflict comparison so "CC0"/"CC-0" spellings are the same value. */
export function normaliseLicenceToken(token: string): string {
  return token.trim().toLowerCase().replace(/[\s_-]+/g, "");
}

/**
 * Permissiveness rank for the more-permissive-wins comparison (redistribution
 * freedom: CC0 imposes no conditions; CC-BY imposes attribution only; copyleft
 * imposes source obligations; an unrecognised token imposes unknown obligations).
 * CC0 / public domain 3, CC-BY family 2, AGPL/GPL copyleft 1, unrecognised 0.
 * Silence has no rank — the callers treat it as "not a candidate", never a winner.
 */
export function licencePermissivenessRank(token: string): number {
  const t = token.trim();
  if (/cc\s*[-_ ]?0\b/i.test(t) || /public\s+domain/i.test(t)) return 3;
  if (/cc[\s_-]*by/i.test(t)) return 2;
  if (/agpl|gpl|copyleft/i.test(t)) return 1;
  return 0;
}
